src/zkevm_specs/evm/execution/callop.py

```diff
@@ -108,21 +108,27 @@ def callop(instruction: Instruction):
     has_value = 1 - instruction.is_zero(value)
     instruction.constrain_zero(has_value * is_static)
 
-    # Verify transfer
-    _, (_, callee_balance_prev) = instruction.transfer(
-        caller_address, callee_address, value, callee_reversion_info
-    )
+    if is_call == 1:
+        # Verify transfer only for CALL opcode.
+        _, (_, callee_balance) = instruction.transfer(
+            caller_address, callee_address, value, callee_reversion_info
+        )
+    else:
+        # Get callee balance for CALLCODE, DELEGATECALL and STATICCALL opcodes.
+        callee_balance = instruction.account_read(callee_address, AccountFieldTag.Balance)
```

Comment:
According to the CALLCODE implementation you linked in https://github.com/privacy-scaling-explorations/zkevm-specs/issues/334#issue-1481042446
For CALLCODE you should constraint that `value <= callee_balance` right?
I think I didn't see this constraint.

Reply:
Fixed to add a constraint to verify caller balance is greater than or equal to stack `value` for both `CALL` and `CALLCODE` opcodes as below:
```
if is_call + is_callcode == 1:
    value_lt_caller_balance, value_eq_caller_balance = instruction.compare_word(
        value, caller_balance
)
instruction.constrain_zero(1 - value_lt_caller_balance - value_eq_caller_balance)
```